File: ioi2020-biscuits/solution/biscuits-yanhao-ac.java
import java.util.HashMap;

class biscuits {
	HashMap<Long, Long> m = new HashMap<>();
	long f(long[] s, long x, long n) {
		if(n<=0) return 0;
		if(n==1) return 1;
		if(m.containsKey(n)) {
			return m.get(n);
		}
		int a = 63-Long.numberOfLeadingZeros(n-1);
		long b = ((long)1)<<a;
		long c = Math.min(n,1+s[a]/x) - b;
		long ans = f(s,x,b) + f(s,x,c);
		m.put(n, f(s,x,b) + f(s,x,c));
		return ans;
	}
	long count_tastiness(long x, long[] a) {
		m = new HashMap<>();
		long[] b = new long[61];
		for(int i=0; i<a.length; i++) {
			b[i] = a[i];
		}
		for(int i=1; i<a.length; i++) {
			b[i] = b[i-1] + (b[i]<<i);
		}
		for(int i=a.length; i<=60; i++) {
			b[i] = b[i-1];
		}
		return f(b, x, 1+b[60]);
	}
}

File: ioi2020-biscuits/solution/biscuits-yanhao-ac.cpp
#include "biscuits.h"
#include <vector>
#include <map>

using namespace std;
map<long long, long long> m;
long long f(vector<long long> &s, long long x, long long n) {
    if(n<=0) return 0;
    if(n==1) return 1;
    if(m.find(n)!=m.end()) {
        return m[n];
    }
    long long a = __lg(n-1);
    return m[n] = f(s,x,1LL<<a) + f(s,x,min(n,1+s[a]/x)-(1LL<<a));
}
long long count_tastiness(long long x, std::vector<long long> a) {
    m.clear();
    for(int i=1; i<(int)a.size(); i++) {
        a[i] = a[i-1] + (a[i]<<i);
    }
    while(a.size()<=60) a.push_back(a.back());
	return f(a, x, 1+a.back());
}
